File: files/9590/lecture10.R
###################################
## Code for POLSCI 9590 Week 9   ##
## Dave Armstrong                ##
## University of Western Ontario ##
## dave.armstrong@uwo.ca         ##
## 2024                          ##
###################################

library(tidyverse)
library(rio)
library(plotly)
library(marginaleffects)

dat = rio::import("example1.dta")
ggplot(dat, 
       mapping=aes(x=x, y=y)) + 
  geom_point(pch=1) +  
  theme_bw() + 
  theme(aspect.ratio=1) 

x <- dat$x
y <- dat$y
a <- seq(1.75, 2.25, length=100)
b <- seq(2.5, 3.5, length=100)
eg <- expand.grid(a=a, b=b)
yhats <- sapply(1:nrow(eg), function(i)eg[i,1] + eg[i,2]*x)
rss <- apply(yhats, 2, function(x)sum((x-y)^2))
eg$rss <- rss
fitmat <- matrix(eg$rss, 
                 nrow=length(a), 
                 ncol=length(b))
plot_ly(x = ~a, y=~b, z=~fitmat) %>% 
  add_surface(colorbar=list(title="RSS"))

x <- dat$x
y <- dat$y
mod <- lm(y ~ x)
g <- ggplot(eg, aes(x=a, y=b, z=rss)) + 
  geom_contour(breaks = quantile(eg$rss, c(.0001, .001, .01, .05, .1,.2,.3,.4,.5,.6)), col="black") + 
  geom_hline(yintercept=mod$coef[2], col="red") +
  geom_vline(xintercept=mod$coef[1], col="red") +
  theme_bw()
gp1 <- ggplotly(g)
gp1

mod <- lm(y ~ x)
summary(mod)

ggplot(dat, aes(x=x, y=y)) + 
  geom_point(shape=1) + 
  geom_smooth(method="lm", 
              col="black") + 
  theme_classic() + 
  labs(x="x", y="y")

demo <- rio::import("demo.dta")
demo <- rio::factorize(demo)
dmod1 <- lm(demodays ~ corrupt, data=demo)
summary(dmod1)

s <- sd(demo$corrupt, na.rm=TRUE)
s*0.32


.79/sd(demo$demodays, na.rm=TRUE)

plot_predictions(dmod1,
                 condition="corrupt", 
                 points=1) + 
  theme_classic()

df <- import("cat_example.dta")
df <- factorize(df)
df %>% group_by(x) %>% 
  summarise(m = mean(y), 
            s = sd(y), 
            n = n())


xmod <- lm(y ~ x, data=df)
summary(xmod)

avg_predictions(xmod, variables="x")
avg_predictions(xmod, variables="x", hypothesis = "pairwise")

plot_predictions(xmod, condition="x") + 
  geom_point(data=df, aes(x=x, y=y), position=position_jitter(width=.15), color="gray50", alpha=.25, inherit.aes = FALSE) + 
  theme_classic()

File: _site/files/9590/lecture10.py
###################################
## Code for POLSCI 9590 Week 9   ##
## Dave Armstrong                ##
## University of Western Ontario ##
## dave.armstrong@uwo.ca         ##
## 2024                          ##
###################################

import pandas as pd
import statsmodels.formula.api as smf
dat = pd.read_stata("example1.dta")
mod_setup =  smf.ols(' y ~ x', data=dat)
mod = mod_setup.fit()

mod.summary(slim=True)

import seaborn as sb
import matplotlib.pyplot as plt
plt.clf()
sb.set_style('ticks')
fig, ax = plt.subplots()
fig.set_size_inches(6, 4)
sb.regplot(x=dat.x, y=dat.y, scatter_kws={"alpha": .1}, ax=ax)
sb.despine()
plt.show()
plt.clf()

demo = pd.read_stata('demo.dta')
mod =  smf.ols('demodays ~ corrupt', data=demo).fit()


mod.summary(slim=True)

plt.clf()
sb.set_style('ticks')
fig, ax = plt.subplots()
fig.set_size_inches(6, 4)
sb.regplot(x=demo.corrupt, y=demo.demodays, scatter_kws={"alpha": .1}, ax=ax)
sb.despine()
plt.show()
plt.clf()

df = pd.read_stata("cat_example.dta")
(df
  .groupby("x", observed=True)
  .agg({'y':['mean', 'std', lambda x: len(x)]})
  .reset_index()
  .droplevel(0, axis=1)
  .rename(columns={'': 'Group', 'mean': 'Mean', 'std': 'SD', '<lambda_0>': 'N'}))


xmod =  smf.ols('y ~ x', data=df).fit()
print(xmod.summary().tables[1].as_text())


import marginaleffects as me
me.avg_predictions(xmod, by="x").to_pandas()
me.avg_predictions(xmod, by="x", hypothesis="pairwise").to_pandas()


from plotnine import geom_point, position_jitter, aes
p = me.plot_predictions(xmod, by="x")  
(p + geom_point(data=df, 
                mapping = aes(x="x", y="y"), 
                position = position_jitter(width=.15), 
                inherit_aes=False)).show()


